reactioncommerce/reaction
# fix: Make new order email links correct for storefront

Resolves #4679   
Impact: **breaking|minor**  
Type: **bugfix**

## Issue

New Order confirmation email was not updated yet to deal with storefront and core being on separate URLs yet.

## Solution

We use the recently-added shop `storefrontUrls` settings to build the URL and pass that to the email template for rendering.

One complication we hit is dealing with access tokens for anonymous orders. We don't store them, only their hash, so while we can navigate to their order confirmation page immediately, later when we go to render their order confirmation email, we don't have the raw token. So we decided to generate a new one at that time which required a schema change so an order can now have an array of access tokens. Typically there will be one from when the order was placed via storefront and one from when the order confirmation email was rendered and sent.

## Breaking changes

The new order confirmation email template has been updated. The new code is here but we also write these to mongodb so they can be edited in the GUI. There is no migration for that data. It would have some extra complexity to determine if the data had been modified or not. Here's the diff if anyone wants to apply manually:

```
-                                              <td width="33%" align="left" valign="top" style="font-size:14px; line-height:normal; color:#1999dd; font-family:Arial, helvetica;"><a href="{{homepage}}{{orderUrl}}" style="color:#1999dd;">{{order.referenceId}}</a></td>
+                                              <td width="33%" align="left" valign="top" style="font-size:14px; line-height:normal; color:#1999dd; font-family:Arial, helvetica;"><a href="{{orderUrl}}" style="color:#1999dd;">{{order.referenceId}}</a></td>
```
A migration was created and tested for the schema change.

## Testing
1. Set up your install for outgoing emails (we have internal docs on this, ping Pete for a link)
1. Create some orders on your existing installation BEFORE loading the new code in this branch. Ideally both logged-in and anonymous orders.
    1. This will give you "before" data to test the migration
1. Configure your Shop Settings in catalyst with
    1. Single Order page URL: `http://storefront.reaction.localhost:4000/checkout/order/:orderId?token=:token`
    1. Homepage URL: `http://storefront.reaction.locahost:4000`
1. Checkout this branch and restart reaction
1. Verify the schema migration is run and you are up to `66` (you should see this in the startup logs for reaction core)
1. Verify the email links to your existing orders still work
1. Place new logged-in and anonymous orders and verify they still work
